### PR TITLE
[NPU] optimization for dsv3.2

### PR DESCRIPTION
**Pull request by ZhengdQin**
Co-author: @jiaming1130

## Motivation

In this PR, we optimized the performance of the DeepSeek-V3.2 on NPU  to improve the performance of the network in the decode phase, and also adapted the dynamic W8A8 quantization for kv_b.

## Modifications

1. Implemented multi-stream parallelism for the shared expert and MOE.
2. Implemented multi-stream parallelism for the calculation of q, k and weight in the indexer.
3. Implemented multi-stream parallelism for the mlapo kernel and fused_qkv_a_proj_weight_mqa (matmul) in the decode phase.
4. Implemented multi-stream parallelism for q_b_proj and kv_a_layernorm in the forward_dsa_prepare_npu function.
5. Optimized the renormalization process after npu_moe_gating_top_k operation.
6. Use forward_native for npu in LayerNorm.
7. Use fusion kernel batch_matmul_transpose in the MTP.
8. Support ascend_fuseep for deepseek-v3 and deepseek-v3.2

## Accuracy Tests

image

## Benchmarking and Profiling

Before:
image

After
image

image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py [hunk 1]
```diff
@@ -358,7 +370,11 @@ def forward_dsa_core_npu(
         device=attn_output.device,
     )
 
-    if not forward_batch.forward_mode.is_decode():
+    if (
+        forward_batch.forward_mode.is_extend()
+        and not forward_batch.forward_mode.is_draft_extend_v2()
+        and not forward_batch.forward_mode.is_target_verify()
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py:385]: ```is_draft_extend(include_v2=True)```

File: python/sglang/srt/hardware_backend/npu/moe/topk.py [hunk 2]
```diff
@@ -39,8 +39,7 @@ def fused_topk_npu(
         topk_weights = topk_weights.to(torch.float32)
 
     elif use_grouped_topk and correction_bias is not None:
-        routed_scaling_factor = topk_config.routed_scaling_factor or 1
-
+        # Force set routed_scaling_factor = 1 to optimize renormalize
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/moe/topk.py:42]: why? what's the impact on accuracy?

File: python/sglang/srt/models/deepseek_nextn.py [hunk 3]
```diff
@@ -97,7 +97,12 @@ def __init__(
 
         self.eh_proj = nn.Linear(2 * config.hidden_size, config.hidden_size, bias=False)
 
-        self.alt_stream = torch.cuda.Stream() if _is_cuda else None
+        if _is_cuda:
+            self.alt_stream = torch.cuda.Stream()
+        elif _is_npu:
+            self.alt_stream = torch.npu.Stream()
+        else:
+            self.alt_stream = None
```
Comment (iforgetmyname) [inline python/sglang/srt/models/deepseek_nextn.py]:
```suggestion
        self.alt_stream = torch.cuda.Stream() if _is_cuda or _is_npu else None
```

File: python/sglang/srt/models/deepseek_v2.py [hunk 4]
```diff
@@ -2972,7 +2984,13 @@ def __init__(
...
             self.embed_tokens = PPMissingLayer()
 
-        self.alt_stream = torch.cuda.Stream() if _is_cuda else None
+        if _is_cuda:
+            self.alt_stream = torch.cuda.Stream()
+        elif _is_npu:
+            self.alt_stream = torch.npu.Stream()
+        else:
+            self.alt_stream = None
+
```
Comment (iforgetmyname) [inline python/sglang/srt/models/deepseek_v2.py]:
```suggestion
        self.alt_stream = torch.cuda.Stream() if _is_cuda or _is_npu else None
```

File: python/sglang/srt/models/deepseek_v2.py [hunk 5]
```diff
@@ -1104,6 +1111,12 @@ def _post_combine_hook(
             topk_output=topk_output,
         )
 
+        if (
+            hidden_states.shape[0] > 0
+            and not self._fuse_shared_experts_inside_sbo
+            and self._use_multi_stream
+        ):
+            torch.get_device_module().current_stream().wait_event(shared_event)
```
Comment (iforgetmyname) [inline python/sglang/srt/models/deepseek_v2.py]: ```torch.cuda.current_stream()```

File: python/sglang/srt/models/deepseek_v2.py [hunk 6]
```diff
@@ -736,6 +733,7 @@ def __init__(
                     )
 
         self.top_k = config.num_experts_per_tok
+        self._use_multi_stream = get_bool_env_var("USE_MULTI_STREAM", "0")
```
Comment (iforgetmyname) [inline python/sglang/srt/models/deepseek_v2.py]: use ```srt/environ.py``` instead
p.s. can we make this default on?

File: python/sglang/srt/layers/layernorm.py [hunk 7]
```diff
@@ -343,19 +343,9 @@ def forward_npu(
...
-
-        if self.elementwise_affine:
-            x = x * self.weight.to(self.dtype)
-            if self.use_bias:
-                x = x + self.bias.to(self.dtype)
-
-        return x.to(orig_dtype)
+        return F.layer_norm(
+            x, (self.hidden_size,), self.weight, self.bias, self.variance_epsilon
+        )
```
Comment (ping1jing2) [inline python/sglang/srt/layers/layernorm.py]: `return self.forward_native(x)`

File: python/sglang/srt/layers/linear.py [hunk 8]
```diff
@@ -245,6 +245,9 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
                 else:
                     raise ValueError(f"{loaded_weight} are not all equal")
 
+            if param.dtype == torch.int8 or loaded_weight.dtype == torch.int8:
+                assert param.dtype == loaded_weight.dtype
```
Comment (ping1jing2) [inline python/sglang/srt/layers/linear.py]: pls add docstring/debug info

File: python/sglang/srt/hardware_backend/npu/quantization/modelslim.py [hunk 9]
```diff
@@ -159,8 +159,10 @@ def get_quant_method(
                     proj_name, packed_modules_mapping_subset[proj_name][0]
                 )
             self.is_dynamic = (
-                self.quant_description[prefix_in_quant_config + ".weight"]
+                self.quant_description.get(prefix_in_quant_config + ".weight", "")
```
Comment (ping1jing2) [inline python/sglang/srt/hardware_backend/npu/quantization/modelslim.py:162]: pls add some comments to explain

File: python/sglang/srt/hardware_backend/npu/moe/topk.py (hunk 2 above)
Comment (ZhengdQin) [reply python/sglang/srt/hardware_backend/npu/moe/topk.py:42]: The kernel npu_moe_gating_top_k can normalize the topk_weights to 1 if we set routed_scaling_factor = 1, so we don't need to re-normalize the topk_weights. Details please see https://www.hiascend.com/doc_center/source/zh/Pytorch/700/apiref/apilist/ptaoplist_002365.html

File: python/sglang/srt/models/deepseek_nextn.py (hunk 3 above)
Comment (ZhengdQin) [reply python/sglang/srt/models/deepseek_nextn.py]: done

File: python/sglang/srt/models/deepseek_v2.py (hunk 4 above)
Comment (ZhengdQin) [reply python/sglang/srt/models/deepseek_v2.py]: done

File: python/sglang/srt/models/deepseek_v2.py (hunk 5 above)
Comment (ZhengdQin) [reply python/sglang/srt/models/deepseek_v2.py]: done

File: python/sglang/srt/models/deepseek_v2.py (hunk 6 above)
Comment (ZhengdQin) [reply python/sglang/srt/models/deepseek_v2.py]: We use the condition `if self.alt_stream is None` to enable multi-stream optimization, which is therefore always enabled for NPU.

File: python/sglang/srt/layers/layernorm.py (hunk 7 above)
Comment (ZhengdQin) [reply python/sglang/srt/layers/layernorm.py]: done

File: python/sglang/srt/layers/linear.py (hunk 8 above)
Comment (ZhengdQin) [reply python/sglang/srt/layers/linear.py]: done

File: python/sglang/srt/hardware_backend/npu/quantization/modelslim.py (hunk 9 above)
Comment (ZhengdQin) [reply python/sglang/srt/hardware_backend/npu/quantization/modelslim.py:162]: done

File: python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py (hunk 1 above)
Comment (ZhengdQin) [reply python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py:385]: done

**Review by ping1jing2 — APPROVED** [review]

**Review by iforgetmyname — APPROVED** [review]
